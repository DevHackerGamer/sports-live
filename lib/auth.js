// Lightweight auth helper that prefers Clerk server verification when available,
// and falls back to a dev header (X-User-Type) if Clerk isn't configured.

let clerkClient = null;
let getAuth = null;
try {
  const clerk = require('@clerk/express');
  clerkClient = clerk.clerkClient;
  getAuth = clerk.getAuth;
} catch (_) {
  // Clerk not installed; dev fallback will be used
}

function getBearer(req) {
  const auth = req.headers['authorization'] || req.headers['Authorization'];
  if (!auth || typeof auth !== 'string') return null;
  const parts = auth.split(' ');
  if (parts.length === 2 && parts[0] === 'Bearer') return parts[1];
  return null;
}

function normalizeRole(meta) {
  if (!meta) return '';
  const type = meta.type || meta.role || '';
  if (Array.isArray(meta.roles) && meta.roles.length) {
    // Prefer explicit admin role if present
    if (meta.roles.map(r => String(r).toLowerCase()).includes('admin')) return 'admin';
  }
  return String(type || '').toLowerCase();
}

// Attempt to resolve user type from Clerk private metadata
async function getUserType(req) {
  // Dev override header
  const devHeaderType = (req.headers['x-user-type'] || '').toString().toLowerCase();
  const devHeaderRole = (req.headers['x-user-role'] || '').toString().toLowerCase();
  const devType = devHeaderType || devHeaderRole;

  // If Clerk middleware/client is not present, return dev
  if (!clerkClient || !getAuth) {
    return devType || '';
  }

  try {
    const { userId } = getAuth(req) || {};
    if (!userId) return devType || '';
  const user = await clerkClient.users.getUser(userId);
  const privateType = normalizeRole(user?.privateMetadata);
  const publicType = normalizeRole(user?.publicMetadata);
  return privateType || publicType || '';
  } catch (e) {
    // On any failure, fall back to dev header
    return devType || '';
  }
}

async function isAdmin(req) {
  const type = await getUserType(req);
  return type === 'admin';
}

module.exports = { getUserType, isAdmin };
