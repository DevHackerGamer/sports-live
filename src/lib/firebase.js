// Firebase client initialization (App + Realtime Database)
import { initializeApp, getApps, getApp } from 'firebase/app';
import { getAnalytics } from 'firebase/analytics';
import {
  getDatabase,
  ref as dbRef,
  update as dbUpdate,
  set as dbSet,
  child as dbChild,
  get as dbGet,
  onValue as dbOnValue  // <-- add this
} from 'firebase/database';

// Firebase web config (public by design)
const firebaseConfig = {
  apiKey: 'AIzaSyDzZ0CAodwCYHgQJahsgcxJrmWiIIbVGM',
  authDomain: 'sports-live-c25ca.firebaseapp.com',
  databaseURL: 'https://sports-live-c25ca-default-rtdb.firebaseio.com',
  projectId: 'sports-live-c25ca',
  storageBucket: 'sports-live-c25ca.firebasestorage.app',
  messagingSenderId: '532399482994',
  appId: '1:532399482994:web:3d1353393886b64dc1baeb',
  measurementId: 'G-8JT6X5DF6B',
};

// Singleton app instance across HMR/dev reloads
const app = getApps().length ? getApp() : initializeApp(firebaseConfig);

// Realtime Database instance
export const db = getDatabase(app);

// Modular functions for convenience
export const ref = dbRef;
export const update = dbUpdate;
export const set = dbSet;
export const child = dbChild;
export const get = dbGet;
export const onValue = dbOnValue;  // <-- export onValue

// Guard analytics for browser-only environments
export const initAnalytics = () => {
  if (typeof window !== 'undefined' && 'measurementId' in firebaseConfig) {
    try {
      return getAnalytics(app);
    } catch (_) {
      // ignore if analytics not supported
    }
  }
  return null;
};

export default app;
