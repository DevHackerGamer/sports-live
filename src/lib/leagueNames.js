// Centralized league name mapping for ESPN league codes
// This provides consistent display names across the entire application

export const LEAGUE_NAMES = {
  // ESPN league codes
  'eng.1': 'Premier League',
  'esp.1': 'La Liga',
  'ita.1': 'Serie A',
  'ger.1': 'Bundesliga',
  'fra.1': 'Ligue 1',
  'uefa.champions': 'Champions League',
  'ned.1': 'Eredivisie',
  'por.1': 'Primeira Liga',
  'eng.2': 'Championship',
  'bra.1': 'Campeonato Brasileiro SÃ©rie A',
  'conmebol.libertadores': 'Copa Libertadores',
  
  // Football-data.org codes (legacy)
  'PL': 'Premier League',
  'PD': 'La Liga',
  'SA': 'Serie A',
  'BL1': 'Bundesliga',
  'FL1': 'Ligue 1',
  'CL': 'Champions League',
  'DED': 'Eredivisie',
  'PPL': 'Primeira Liga',
  'ELC': 'Championship'
};

/**
 * Get display name for a league code
 * @param {string} code - ESPN league code (e.g., 'eng.1') or name with bracket notation (e.g., 'Bundesliga [ger.1]')
 * @returns {string} Display name (e.g., 'Premier League')
 */
export function getLeagueName(code) {
  if (!code) return 'Unknown League';
  
  // Handle strings with bracket notation like "Bundesliga [ger.1]"
  const bracketMatch = /^(.*?)\s*\[([^\]]+)\]$/.exec(String(code));
  if (bracketMatch) {
    const [, namePart, codePart] = bracketMatch;
    // Try to get the name from the code part first
    const nameFromCode = LEAGUE_NAMES[codePart];
    if (nameFromCode) return nameFromCode;
    // Otherwise return the name part without brackets
    return namePart.trim();
  }
  
  return LEAGUE_NAMES[code] || code;
}

/**
 * Get league code from display name or return as-is if already a code
 * @param {string} nameOrCode - League name or code
 * @returns {string} League code
 */
export function getLeagueCode(nameOrCode) {
  if (!nameOrCode) return '';
  
  // If it's already a code, return it
  if (LEAGUE_NAMES[nameOrCode]) return nameOrCode;
  
  // Find code by name
  const entry = Object.entries(LEAGUE_NAMES).find(([_, name]) => name === nameOrCode);
  return entry ? entry[0] : nameOrCode;
}

/**
 * Format competition object to ensure it has proper display name
 * @param {Object} competition - Competition object with name/code
 * @returns {Object} Competition with displayName
 */
export function formatCompetition(competition) {
  if (!competition) return { name: 'Unknown League', displayName: 'Unknown League' };
  
  const name = competition.name || competition.code || competition.id || '';
  const displayName = getLeagueName(name);
  
  return {
    ...competition,
    displayName,
    code: competition.code || name
  };
}
