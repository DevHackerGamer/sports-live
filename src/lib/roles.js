// Client-side role utilities to interpret Clerk metadata consistently

export function getUserRoles(user) {
  if (!user) return [];
  const raw = [
    user?.privateMetadata?.type,
    user?.publicMetadata?.type,

  ]
    .concat(Array.isArray(user?.privateMetadata?.roles) ? user.privateMetadata.roles : [])
    .concat(Array.isArray(user?.publicMetadata?.roles) ? user.publicMetadata.roles : [])
    .filter(Boolean)
    .map(v => String(v).toLowerCase());
  return Array.from(new Set(raw));
}

export function isAdminFromUser(user) {
  const roles = getUserRoles(user);
  return roles.includes('admin');
}
