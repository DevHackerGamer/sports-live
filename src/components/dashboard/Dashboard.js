import React, { useEffect, useState } from 'react';
import { useUser, useAuth, UserButton } from '@clerk/clerk-react';
import LiveSports from '../sports/LiveSports';
import FavoritesPanel from '../favouritespanel/FavoritesPanel';
import WatchlistPage from '../WatchlistPage/WatchlistPage';
import MatchViewer from '../matchViewer/MatchViewer';
import MatchSetup from '../matchsetup/MatchSetup';
import LiveInput from '../liveInput/LiveInput';
import LeagueView from '../LeagueView/LeagueView';
import ReportsPage from '../ReportsPage/ReportsPage';
import PlayersPage from '../PlayersPage/PlayersPage';
import TeamInfo from '../TeamInfo/TeamInfo';
import { isAdminFromUser, getUserRoles } from '../../lib/roles';

// Import league images
import plLogo from '../../assets/pl_logo.jpg';
import laLigaLogo from '../../assets/LaLiga.jpg';
import serieALogo from '../../assets/serie_A.jpg';
import bundesligaLogo from '../../assets/bundesliga.jpg';
import ligue1Logo from '../../assets/LIGUE1.jpg';
import championsLogo from '../../assets/UCL.jpg';

import '../../styles/Dashboard.css';


//refreshing page logic usin states
// --- Dashboard State Persistence Helpers ---
const DASHBOARD_STATE_KEY = "sportslive-dashboard-state";

const saveDashboardState = (state) => {
  try {
    localStorage.setItem(DASHBOARD_STATE_KEY, JSON.stringify(state));
  } catch (err) {
    console.error("Error saving dashboard state:", err);
  }
};

const loadDashboardState = () => {
  try {
    const saved = localStorage.getItem(DASHBOARD_STATE_KEY);
    return saved ? JSON.parse(saved) : null;
  } catch (err) {
    console.error("Error loading dashboard state:", err);
    return null;
  }
};


const Dashboard = () => {
  const { user } = useUser();
  const { getToken, isSignedIn } = useAuth();
  const clerkIsAdmin = isAdminFromUser(user);
  const [isAdmin, setIsAdmin] = useState(!!clerkIsAdmin);
  const [activeTab, setActiveTab] = useState('home');
  const [selectedMatch, setSelectedMatch] = useState(null);
  const [showAboutUs, setShowAboutUs] = useState(false);
  const [selectedLeague, setSelectedLeague] = useState(null);
  const [selectedTeam, setSelectedTeam] = useState(null); 
  const [selectedTabTitle, setSelectedTabTitle] = useState('home');

  // Sync isAdmin with Clerk
  useEffect(() => {
    setIsAdmin(isAdminFromUser(user));
  }, [user]);

  //refreshing page logic usin states
// --- Dashboard State Persistence Helpers ---
useEffect(() => {
  const saved = loadDashboardState();
  if (saved) {
    if (saved.activeTab) setActiveTab(saved.activeTab);
    if (saved.selectedMatch) setSelectedMatch(saved.selectedMatch);
    if (saved.selectedTeam) setSelectedTeam(saved.selectedTeam);
    if (saved.selectedLeague) setSelectedLeague(saved.selectedLeague);
    if (saved.showAboutUs !== undefined) setShowAboutUs(saved.showAboutUs);
  }
}, []);

useEffect(() => {
  saveDashboardState({
    activeTab,
    selectedMatch,
    selectedTeam,
    selectedLeague,
    showAboutUs,
  });
}, [activeTab, selectedMatch, selectedTeam, selectedLeague, showAboutUs]);





  // Resolve role from backend
  useEffect(() => {
    let cancelled = false;
    (async () => {
      try {
        const headers = new Headers();
        try {
          const token = await getToken();
          if (token) headers.set('Authorization', `Bearer ${token}`);
        } catch {}
        const userRoles = getUserRoles(user);
        const devType = (userRoles[0] || '').toString();
        const roleHeader = userRoles.join(',');
        if (devType) headers.set('X-User-Type', devType);
        if (roleHeader) headers.set('X-User-Role', roleHeader);

        const res = await fetch('/api/auth-me', { headers });
        const data = await res.json().catch(() => ({}));
        if (!cancelled) setIsAdmin(prev => !!prev || !!data.isAdmin);
      } catch (e) {
        if (!cancelled) setIsAdmin(prev => !!prev);
      }
    })();
    return () => { cancelled = true; };
  }, [user, isSignedIn, getToken]);

  const handleMatchSelect = (match) => setSelectedMatch(match);
  const handleBackFromViewer = () => setSelectedMatch(null);
  const handleTeamSelect = (team) => setSelectedTeam(team); 
  const handleBackFromTeamInfo = () => setSelectedTeam(null);

  // Watchlist now handled via LiveSports hover UI and user-matches API.

  // League Data
  const leagues = [
    {
      key: "PL",
      name: "Premier League",
      img: plLogo,
      desc: "The most-watched league in the world, known for its relentless pace, physicality, and electrifying atmospheres.",
      color: "#331160ff"
    },
    {
      key: "LL",
      name: "La Liga",
      img: laLigaLogo,
      desc: "Home of technical mastery and tactical artistry, where legendary rivalries and world-class talent take center stage.",
      color: "Red"
    },
    {
      key: "SA",
      name: "Serie A",
      img: serieALogo,
      desc: "The birthplace of 'Catenaccio,' renowned for its tactical sophistication, defensive rigor, and storied history.",
      color: "#0c4491"
    },
    {
      key: "BL",
      name: "Bundesliga",
      desc: "A league famous for its passionate fan culture, high-scoring games, and the development of world-class young talent.",
      img: bundesligaLogo,
      color: "#d71218"
    },
    {
      key: "L1",
      name: "Ligue 1",
      img: ligue1Logo,
      desc: "A rapidly growing league celebrated for its exciting blend of emerging superstars and established international stars.",
      color: "#0055A4"
    },
    {
      key: "UCL",
      name: "Champions League",
      img: championsLogo,
      desc: "The ultimate stage in club football, where European giants collide in pursuit of the most coveted trophy in the sport.",
      color: "#272727"
    },
  ];

  // Home screen
  const HomeScreen = () => {
    const features = [
      { 
        title: "Live Match Updates", 
        description: "Real-time scores and updates", 
        icon: "⚽",
        action: () => setActiveTab('matches')
      },
      { 
        title: "Player Statistics", 
        description: "Detailed stats for all players", 
        icon: "👤",
        action: () => setActiveTab('players')
      },
      { 
        title: "Team Standings", 
        description: "Track team positions in league tables", 
        icon: "🏆",
        action: () => setActiveTab('leagueStandings')
      },
      { 
        title: "Personalized Experience", 
        description: "Follow your favorite teams and players", 
        icon: "❤️",
        action: () => setActiveTab('favorites')
      }
    ];

    return (
      <div className="home-screen">
        {/* Hero Section - Top part */}
        <section className="hero">
          <div className="hero-content">
            <h1>Experience Football <span>Like Never Before</span></h1>
            <p>Real-time scores, in-depth statistics, and coverage of all major leagues.</p>
          </div>
        </section>

        {/* Features */}
        <section className="features-section">
          <h2 className="section-title">Why Choose Sports Live?</h2>
          <p className="section-subtitle">Everything you need to stay connected to the world of football</p>
          <div className="features-grid">
            {features.map((f, idx) => (
              <div className="feature-card" key={idx} onClick={f.action}>
                <div className="feature-icon">{f.icon}</div>
                <h3>{f.title}</h3>
                <p>{f.description}</p>
              </div>
            ))}
          </div>
        </section>

        {/* League Cards */}
        <section className="league-cards">
          <h2 className="section-title">Top Leagues</h2>
          <p className="section-subtitle">Follow your favorite leagues with comprehensive coverage</p>
          <div className="cards-grid">
            {leagues.map((league, idx) => (
              <div className="league-card" key={idx}>
                <div className="league-card-inner">
                  <div className="league-card-front">
                    <div className="league-img-container">
                      <img src={league.img} alt={league.name} className="league-img" />
                    </div>
                  </div>
                  <div className="league-card-back" style={{ backgroundColor: league.color }}>
                    <div className="league-card-content">
                      <p>{league.desc}</p>
                      <button 
                        className="btn btn-small"
                        onClick={() => {
                          setSelectedLeague(league.key);
                          setActiveTab('leagueStandings');
                        }}
                      >
                        View {league.name}
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            ))}
          </div>
        </section>
      </div>
    );
  };

  // About Us Section 
  const AboutUs = () => (
    <div className="about-us-container">
      <div className="container">
        <h2>About Sports Live</h2>
        <div className="about-content">
          <p>Sports Live is the premier destination for football enthusiasts who want real-time updates, comprehensive statistics, and in-depth coverage of all major football leagues around the world.</p>
          <h3>Our Mission</h3>
          <p>To provide football fans with the most accurate, timely, and comprehensive football data and match coverage in an intuitive and engaging platform.</p>
          <h3>What We Offer</h3>
          <ul>
            <li>Real-time match updates and live scores</li>
            <li>Detailed player and team statistics</li>
            <li>League standings and tournament progress</li>
            <li>Personalized favorites system to follow your preferred teams</li>
            <li>Comprehensive coverage of Premier League, La Liga, Serie A, Bundesliga, Ligue 1, and European competitions</li>
          </ul>
          <h3>Our Team</h3>
          <p>We are a passionate group of football enthusiasts, data analysts, and developers dedicated to creating the best football experience for fans worldwide.</p>
          <h3>Contact Us</h3>
          <p>Have questions or feedback? Reach out to us at support@sportslive.com</p>
        </div>
        <button 
          className="btn btn-secondary mt-3" 
          onClick={() => {
            setShowAboutUs(false);
            setActiveTab('home');
          }}
        >
          Back to Dashboard
        </button>
      </div>
    </div>
  );

  // Footer Component
  const Footer = () => (
    <footer className="site-footer">
      <div className="footer-container">
        <div className="footer-section">
          <h3>Sports Live</h3>
          <p>Your ultimate destination for real-time football updates, statistics, and league coverage.</p>
        </div>
        
        <div className="footer-section">
          <h3>Quick Links</h3>
          <ul className="footer-quick-links">
            <li><button onClick={() => { setActiveTab('home'); setShowAboutUs(false); setSelectedMatch(null); setSelectedTeam(null); }}>Home</button></li>
            <li><button onClick={() => { setActiveTab('matches'); setShowAboutUs(false); setSelectedMatch(null); setSelectedTeam(null); }}>Matches</button></li>
            <li><button onClick={() => { setActiveTab('favorites'); setShowAboutUs(false); setSelectedMatch(null); setSelectedTeam(null); }}>Favorites</button></li>
            <li><button onClick={() => { setActiveTab('players'); setShowAboutUs(false); setSelectedMatch(null); setSelectedTeam(null); }}>Players</button></li>
          </ul>
        </div>
        
        <div className="footer-section">
          <h3>Leagues</h3>
          <ul className="footer-quick-links">
            {leagues.map(l => (
              <li key={l.key}>
                <button onClick={() => {
                  setSelectedLeague(l.key);
                  setActiveTab('leagueStandings');
                  setShowAboutUs(false);
                  setSelectedMatch(null);
                  setSelectedTeam(null);
                }}>
                  {l.name}
                </button>
              </li>
            ))}
          </ul>
        </div>
        
        <div className="footer-section">
          <h3>Newsletter</h3>
          <p>Subscribe to get updates on new features and matches</p>
          <div className="newsletter-form">
            <input type="email" placeholder="Your email address" />
            <button>Subscribe</button>
          </div>
        </div>
      </div>
      
      <div className="footer-bottom">
        <p>&copy; {new Date().getFullYear()} Sports Live. All rights reserved.</p>
      </div>
    </footer>
  );

const renderContent = () => {
  if (showAboutUs) return <AboutUs />;

    if (selectedTeam) {
    return (
      <TeamInfo 
        team={selectedTeam} 
        onBack={handleBackFromTeamInfo} 
      />
    );
  }

  // from match cards
  const handleMatchSelect = (match) => {
  setSelectedMatch(match);
  setActiveTab('matches'); // 
};

const handleBackFromViewer = () => {
  setSelectedMatch(null);
  setActiveTab('leagueStandings');
};





  switch (activeTab) {
    case 'home':
      return <HomeScreen />;
    case 'players':
      return <PlayersPage />;
    case 'matchSetup':
      return <MatchSetup isAdmin={isAdmin} />;
    case 'liveInput':
      return <LiveInput isAdmin={isAdmin} match={selectedMatch} onBackToMatch={() => setActiveTab('matches')} />;
    case 'leagueStandings':
      return <LeagueView initialLeague={selectedLeague || "PL"} onBack={() => setActiveTab('home')} onTeamSelect={handleTeamSelect} onMatchSelect={handleMatchSelect} />;
    case 'reports':
      return <ReportsPage isAdmin={isAdmin} />;
    case 'favorites':
      return <FavoritesPanel onMatchSelect={handleMatchSelect} />;
    case 'watchlist':
      return <WatchlistPage onMatchSelect={handleMatchSelect} />;
    case 'matches':
    default:
      return selectedMatch ? (
        <MatchViewer
          match={selectedMatch}
          onBack={handleBackFromViewer}
        />
      ) : (
        <LiveSports onMatchSelect={handleMatchSelect} onTeamSelect={handleTeamSelect} />
      );
  }
};


  return (
    <div className="site-wrap dashboard">
      <header className="dashboard-header">
        <div className="header-left">
          <div className="logo-trapezoid">
            <span className="logo-text">SportsLive</span>
          </div>
          <nav className="dashboard-nav">
            <ul>
              <li><button onClick={() => { setActiveTab('home'); setShowAboutUs(false); setSelectedTeam(null); }}>Home</button></li>
              <li><button onClick={() => { setActiveTab('favorites'); setShowAboutUs(false); setSelectedMatch(null); setSelectedTeam(null); }}>Favourites</button></li>
              <li><button onClick={() => { setActiveTab('watchlist'); setShowAboutUs(false); setSelectedMatch(null); setSelectedTeam(null); }}>Watchlist</button></li>
              {isAdmin && (
                <>
                  <li><button onClick={() => { setActiveTab('matchSetup'); setShowAboutUs(false);  setSelectedTeam(null); }}>Setup</button></li>
                  {selectedMatch && (
                    <li>
                      <button
                        onClick={() => {
                          setActiveTab('liveInput');
                          setShowAboutUs(false);
                          setSelectedTeam(null);
                        }}
                        title="Enter live input for the selected match"
                      >
                        Live Input
                      </button>
                    </li>
                  )}
                  <li><button onClick={() => { setActiveTab('reports'); setShowAboutUs(false); setSelectedMatch(null); setSelectedTeam(null); }}>Reports</button></li>
                </>
              )}
              <li><button onClick={() => { setShowAboutUs(true); setActiveTab(''); setSelectedMatch(null); setSelectedTeam(null); }}>About</button></li>
            </ul>
          </nav>
        </div>
        <div className="header-right">
          <div className="user-section">
            <span>Welcome, {user?.firstName || 'User'}</span>
            <UserButton />
          </div>
        </div>
      </header>
      <main className="site-main">{renderContent()}</main>
      <Footer />
    </div>
  );
};

export default Dashboard;