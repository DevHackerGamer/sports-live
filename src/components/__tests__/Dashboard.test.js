import React from 'react';
import { render, screen, fireEvent, waitFor, act } from '@testing-library/react';
import Dashboard from '../dashboard/Dashboard';
import * as ClerkReact from '@clerk/clerk-react';
import * as roles from '../../lib/roles';

// Mock Clerk hooks and components
jest.mock('@clerk/clerk-react', () => ({
  useUser: jest.fn(),
  useAuth: jest.fn(),
}));

// Mock child components
jest.mock('../../components/Header/Header', () => ({ 
  activeTab, setActiveTab, setShowAboutUs, setSelectedMatch, setSelectedTeam, isAdmin, selectedMatch, selectedMatchId 
}) => (
  <div data-testid="header">
    Header - Active: {activeTab} - Admin: {isAdmin ? 'Yes' : 'No'}
    <button onClick={() => setActiveTab('matches')}>Go to Matches</button>
    <button onClick={() => setShowAboutUs(true)}>Show About</button>
  </div>
));

jest.mock('../../components/Footer/Footer', () => ({ 
  setActiveTab, setShowAboutUs, setSelectedMatch, setSelectedTeam, leagues, setSelectedLeague 
}) => (
  <div data-testid="footer">
    Footer - Leagues: {leagues.length}
    <button onClick={() => setActiveTab('home')}>Go Home</button>
  </div>
));

jest.mock('../../components/MainContent/MainContent', () => ({ 
  showAboutUs, selectedTeam, selectedMatch, activeTab, isAdmin, setActiveTab, setShowAboutUs, setSelectedMatch 
}) => (
  <div data-testid="main-content">
    MainContent - Tab: {activeTab} - ShowAbout: {showAboutUs ? 'Yes' : 'No'}
    {selectedTeam && <span>Team: {selectedTeam.name}</span>}
    {selectedMatch && <span>Match: {selectedMatch.id}</span>}
    <button onClick={() => setActiveTab('favorites')}>Go to Favorites</button>
  </div>
));

jest.mock('../../components/HomeScreen/HomeScreen', () => ({ 
  setActiveTab, setSelectedLeague, leagues, latestNews, newsLoading 
}) => (
  <div data-testid="home-screen">
    HomeScreen - News: {latestNews.length} - Loading: {newsLoading ? 'Yes' : 'No'}
    <button onClick={() => setActiveTab('leagueStandings')}>View Standings</button>
  </div>
));

jest.mock('../../components/AboutUs/AboutUs', () => ({ setShowAboutUs, setActiveTab }) => (
  <div data-testid="about-us">
    About Us
    <button onClick={() => { setShowAboutUs(false); setActiveTab('home'); }}>Back to Dashboard</button>
  </div>
));

jest.mock('../HighlightsTab/HighlightsTab', () => () => (
  <div data-testid="highlights-tab">Highlights Tab</div>
));

// Mock window location and history
const mockPushState = jest.fn();
const mockAddEventListener = jest.fn();
const mockRemoveEventListener = jest.fn();

Object.defineProperty(window, 'location', {
  value: {
    pathname: '/dashboard',
    search: '',
  },
  writable: true,
});

Object.defineProperty(window, 'history', {
  value: {
    pushState: mockPushState,
  },
  writable: true,
});

Object.defineProperty(window, 'addEventListener', {
  value: mockAddEventListener,
  writable: true,
});

Object.defineProperty(window, 'removeEventListener', {
  value: mockRemoveEventListener,
  writable: true,
});

describe('Dashboard Component', () => {
  const mockUser = { 
    id: 'user123', 
    firstName: 'John',
    lastName: 'Doe'
  };
  
  const mockGetToken = jest.fn();
  const mockIsSignedIn = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    
    // Default Clerk mocks
    ClerkReact.useUser.mockReturnValue({ user: mockUser });
    ClerkReact.useAuth.mockReturnValue({ 
      getToken: mockGetToken,
      isSignedIn: jest.fn().mockReturnValue(true)
    });
    
    // Default role mock
    jest.spyOn(roles, 'isAdminFromUser').mockReturnValue(false);
    jest.spyOn(roles, 'getUserRoles').mockReturnValue(['user']);
    
    // Reset window location
    window.location.pathname = '/dashboard';
    window.location.search = '';
    
    // Reset fetch mock
    global.fetch = jest.fn();
  });

  it('renders dashboard with all main components', () => {
    render(<Dashboard />);

    expect(screen.getByTestId('header')).toBeInTheDocument();
    expect(screen.getByTestId('main-content')).toBeInTheDocument();
    expect(screen.getByTestId('footer')).toBeInTheDocument();
  });

  it('initializes with home tab by default', () => {
    render(<Dashboard />);

    expect(screen.getByTestId('main-content')).toHaveTextContent('Tab: home');
  });

  it('handles route-based tab initialization', () => {
    window.location.pathname = '/dashboard/favorites';
    render(<Dashboard />);

    expect(screen.getByTestId('main-content')).toHaveTextContent('Tab: favorites');
  });

  it('handles match ID in URL', () => {
    window.location.pathname = '/dashboard/matches/match123';
    render(<Dashboard />);

    // Should set selectedMatchId from URL
    expect(screen.getByTestId('main-content')).toBeInTheDocument();
  });

  it('handles league parameter in URL', () => {
    window.location.search = '?league=PL';
    render(<Dashboard />);

    // Should set selectedLeague from URL parameter
    expect(screen.getByTestId('main-content')).toBeInTheDocument();
  });

  it('sets admin status from Clerk user', () => {
    jest.spyOn(roles, 'isAdminFromUser').mockReturnValue(true);
    
    render(<Dashboard />);

    expect(screen.getByTestId('header')).toHaveTextContent('Admin: Yes');
  });

  it('fetches admin status from backend API', async () => {
    mockGetToken.mockResolvedValue('mock-token');
    global.fetch.mockResolvedValue({
      ok: true,
      json: () => Promise.resolve({ isAdmin: true })
    });

    await act(async () => {
      render(<Dashboard />);
    });

    await waitFor(() => {
      expect(global.fetch).toHaveBeenCalledWith('/api/auth-me', expect.any(Object));
    });
  });

  it('handles match selection and updates URL', async () => {
    render(<Dashboard />);

    const mockMatch = { id: 'match456', homeTeam: { name: 'Team A' } };
    
    // We need to access the handleMatchSelect function from MainContent
    // This would typically be tested through integration, but for unit test we'll simulate
    const mainContent = screen.getByTestId('main-content');
    
    // The actual URL update would happen in the callback
    // For unit testing, we verify the function exists and is passed down
    expect(mainContent).toBeInTheDocument();
  });

  it('fetches latest news when home tab is active', async () => {
    const mockNews = [
      { _id: '1', headline: 'Test News', description: 'Test description', published: '2024-01-15T10:00:00Z' }
    ];
    
    global.fetch.mockResolvedValue({
      ok: true,
      json: () => Promise.resolve(mockNews)
    });

    await act(async () => {
      render(<Dashboard />);
    });

    await waitFor(() => {
      expect(global.fetch).toHaveBeenCalledWith('/api/football-news?limit=6');
    });
  });

  it('handles back navigation from match viewer', () => {
    render(<Dashboard />);

    // This would typically be tested by simulating the back action
    // For unit testing, we verify the handler function exists
    const mainContent = screen.getByTestId('main-content');
    expect(mainContent).toBeInTheDocument();
  });

  it('handles back navigation from team info', () => {
    render(<Dashboard />);

    // This would typically be tested by simulating the back action
    // For unit testing, we verify the handler function exists
    const mainContent = screen.getByTestId('main-content');
    expect(mainContent).toBeInTheDocument();
  });

  it('updates URL when changing tabs', () => {
    render(<Dashboard />);

    // Simulate tab change through Header component
    const goToMatchesButton = screen.getByText('Go to Matches');
    fireEvent.click(goToMatchesButton);

    expect(mockPushState).toHaveBeenCalled();
  });

  it('handles popstate events for browser navigation', () => {
    render(<Dashboard />);

    expect(mockAddEventListener).toHaveBeenCalledWith('popstate', expect.any(Function));
  });

  it('cleans up event listeners on unmount', () => {
    const { unmount } = render(<Dashboard />);

    unmount();

    expect(mockRemoveEventListener).toHaveBeenCalledWith('popstate', expect.any(Function));
  });

  it('handles user without token gracefully', async () => {
    mockGetToken.mockRejectedValue(new Error('No token'));
    global.fetch.mockResolvedValue({
      ok: true,
      json: () => Promise.resolve({})
    });

    await act(async () => {
      render(<Dashboard />);
    });

    // Should still render without errors
    expect(screen.getByTestId('header')).toBeInTheDocument();
  });

  it('handles admin API fetch errors gracefully', async () => {
    mockGetToken.mockResolvedValue('mock-token');
    global.fetch.mockRejectedValue(new Error('API error'));

    await act(async () => {
      render(<Dashboard />);
    });

    // Should still render without errors
    expect(screen.getByTestId('header')).toBeInTheDocument();
  });

  it('provides league data to child components', () => {
    render(<Dashboard />);

    expect(screen.getByTestId('footer')).toHaveTextContent('Leagues: 6');
  });

  it('handles about us navigation', () => {
    render(<Dashboard />);

    const showAboutButton = screen.getByText('Show About');
    fireEvent.click(showAboutButton);

    expect(mockPushState).toHaveBeenCalledWith({}, '', '/dashboard/about');
  });

  it('matches snapshot with default state', () => {
    const { container } = render(<Dashboard />);
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with admin user', () => {
    jest.spyOn(roles, 'isAdminFromUser').mockReturnValue(true);
    
    const { container } = render(<Dashboard />);
    expect(container).toMatchSnapshot();
  });

  describe('URL parameter handling', () => {
    it('handles different route tabs', () => {
      window.location.pathname = '/dashboard/watchlist';
      render(<Dashboard />);

      expect(screen.getByTestId('main-content')).toHaveTextContent('Tab: watchlist');
    });

    it('handles about route', () => {
      window.location.pathname = '/dashboard/about';
      render(<Dashboard />);

      expect(screen.getByTestId('main-content')).toHaveTextContent('Tab: about');
    });

    it('handles nested match routes', () => {
      window.location.pathname = '/dashboard/matches/match789';
      render(<Dashboard />);

      expect(screen.getByTestId('main-content')).toBeInTheDocument();
    });
  });

  describe('News fetching', () => {
    it('only fetches news when home tab is active', async () => {
      window.location.pathname = '/dashboard/favorites';
      global.fetch.mockResolvedValue({
        ok: true,
        json: () => Promise.resolve([])
      });

      await act(async () => {
        render(<Dashboard />);
      });

      expect(global.fetch).not.toHaveBeenCalledWith('/api/football-news?limit=6');
    });

    it('handles news fetch errors', async () => {
      global.fetch.mockRejectedValue(new Error('News fetch failed'));

      await act(async () => {
        render(<Dashboard />);
      });

      // Should not crash, just log error
      expect(screen.getByTestId('main-content')).toBeInTheDocument();
    });
  });
});