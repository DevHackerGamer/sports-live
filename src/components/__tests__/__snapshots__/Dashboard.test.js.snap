// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Component matches snapshot with admin user 1`] = `
<div>
  <div
    class="site-wrap dashboard"
  >
    <div
      data-testid="header"
    >
      Header - Active: 
      home
       - Admin: 
      Yes
      <button>
        Go to Matches
      </button>
      <button>
        Show About
      </button>
    </div>
    <div
      data-testid="main-content"
    >
      MainContent - Tab: 
      home
       - ShowAbout: 
      No
      <button>
        Go to Favorites
      </button>
    </div>
    <div
      data-testid="footer"
    >
      Footer - Leagues: 
      6
      <button>
        Go Home
      </button>
    </div>
  </div>
</div>
`;

exports[`Dashboard Component matches snapshot with default state 1`] = `
<div>
  <div
    class="site-wrap dashboard"
  >
    <div
      data-testid="header"
    >
      Header - Active: 
      home
       - Admin: 
      No
      <button>
        Go to Matches
      </button>
      <button>
        Show About
      </button>
    </div>
    <div
      data-testid="main-content"
    >
      MainContent - Tab: 
      home
       - ShowAbout: 
      No
      <button>
        Go to Favorites
      </button>
    </div>
    <div
      data-testid="footer"
    >
      Footer - Leagues: 
      6
      <button>
        Go Home
      </button>
    </div>
  </div>
</div>
`;
